spec/lib/extensions/ar_migration_spec.rb

```diff
@@ -0,0 +1,110 @@
+  end
+end
+
+shared_context "with a dummy version" do
+  let(:version) { "1234567890" }
+
+  # sanity check - if this is somehow a version we have, these tests will make no sense
+  before { expect(my_region.migrations_ran).not_to include(version) }
+end
+
+context "with a region seeded" do
+  let!(:my_region) do
+    MiqRegion.seed
+    MiqRegion.my_region
+  end
+
+  describe ArPglogicalMigration::PglogicalMigrationHelper do
+    context "without the migrations ran column" do
+      include_context "without the migrations ran column"
+
+      describe ".migrations_column_present?" do
+        it "is falsey" do
+          expect(described_class.migrations_column_present?).to be_falsey
+        end
+      end
+
+      describe ".update_local_migrations_ran" do
+        it "does nothing" do
+          expect(MiqRegion).not_to receive(:my_region)
+          described_class.update_local_migrations_ran("12345", :up)
+        end
+      end
+    end
+
+    describe ".migrations_column_present?" do
+      it "is truthy" do
+        # we never want to remove this column so we can just test directly
+        expect(described_class.migrations_column_present?).to be_truthy
+      end
+    end
+
+    describe ".update_local_migrations_ran" do
+      include_context "with a dummy version"
+
+      it "adds the given version when the direction is :up" do
+        described_class.update_local_migrations_ran(version, :up)
+        expect(my_region.reload.migrations_ran).to match_array(ActiveRecord::SchemaMigration.normalized_versions << version)
+      end
+
+      it "doesn't blow up when there is no region" do
+        MiqRegion.destroy_all
+        MiqRegion.my_region_clear_cache
+        described_class.update_local_migrations_ran(version, :up)
+      end
+    end
+  end
+
+  describe ArPglogicalMigration::RemoteRegionMigrationWatcher do
+    include_context "with a dummy version"
+
+    let(:subscription) { double("Subscription", :enable => nil, :disable => nil, :provider_region => my_region.region) }
+
+    subject do
+      described_class.new(subscription, version).tap do |s|
+        allow(s).to receive_messages(:puts => nil, :print => nil)
+      end
+    end
+
+    describe "#wait_for_remote_region_migrations" do
+      context "without the migrations ran column present" do
+        include_context "without the migrations ran column"
+
+        it "does nothing" do
+          expect(Vmdb.rails_logger).not_to receive(:info)
+          subject.wait_for_remote_region_migration
+        end
+      end
+
+      it "sleeps until the migration is added" do
+        allow(subject).to receive(:restart_subscription)
+        allow(subject.region).to receive(:reload)
+
+        subject.region.update_attributes!(:migrations_ran => nil)
+
+        t = Thread.new do
+          Thread.current.abort_on_exception = true
+          subject.wait_for_remote_region_migration(0)
+        end
```

Comment:
This test scares me a little.  It looks like it should be fine but threading is hard.

Reply:
I'm fine with merging it as is but it does make me 😨